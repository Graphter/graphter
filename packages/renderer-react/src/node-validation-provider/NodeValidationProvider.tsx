import React, { createContext, useContext } from 'react';
import { NodeConfig, ValidationResult } from "microo-core";

interface DataProviderProps {
  instanceId: string | number,
  nodeValidationHook: NodeValidationHook,
  config: NodeConfig,
  children: any
}

export interface NodeValidationHook {
  (
    instanceId: string | number,
    config: NodeConfig,
    originalNodeData: any,
    storeId?: string,
  ): Array<ValidationResult>
}

const Context = createContext<{ instanceId: string | number, nodeValidationHook: NodeValidationHook } | null>(null);

export const useNodeValidation = (
  config: NodeConfig,
  originalNodeData: any,
  storeId?: string
): Array<ValidationResult> => {
  const ctx = useContext(Context);
  if (!ctx || !ctx.nodeValidationHook) throw new Error(`Couldn't find a NodeValidationHook to use.`);
  return ctx.nodeValidationHook(ctx.instanceId, config, originalNodeData, storeId);
}

export default function NodeValidationProvider(
  {
    instanceId,
    nodeValidationHook,
    children
  }: DataProviderProps
) {
  return (
    <Context.Provider value={{ instanceId, nodeValidationHook }}>
      {children}
    </Context.Provider>
  );
}