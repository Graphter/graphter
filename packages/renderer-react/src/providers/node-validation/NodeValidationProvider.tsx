import React, { createContext, useContext } from 'react';
import { NodeConfig, NodeValidatorRegistration, PathSegment } from "@graphter/core";
import { NodeValidationHook } from "./NodeValidationHook";
import { AggregateNodeValidationHook } from "./AggregateNodeValidationHook";
import { NodeValidationData } from "./NodeValidationData";

interface ValidationProviderProps {
  nodeValidationHook: NodeValidationHook,
  aggregateNodeValidationHook: AggregateNodeValidationHook,
  validatorRegistry: Array<NodeValidatorRegistration>
  children: any
}

const Context = createContext<{
  nodeValidationHook: NodeValidationHook,
  aggregateNodeValidationHook: AggregateNodeValidationHook,
  validatorRegistry: Array<NodeValidatorRegistration>
} | null>(null);

export const useNodeValidation = (
  config: NodeConfig,
  path: Array<PathSegment>
): NodeValidationData => {
  const ctx = useContext(Context);
  if(!ctx) throw new Error('No validation provider found. Make sure a <NodeValidationProvider /> is defined.')
  if (!ctx.nodeValidationHook) throw new Error(`Couldn't find a NodeValidationHook to use.`);
  if(!ctx.validatorRegistry) throw new Error('No validators have been defined. Please ensue <NodeValidationProvider /> has been defined correctly.')

  return ctx.nodeValidationHook(config, path, ctx.validatorRegistry);
}

export const useAggregateNodeValidation = (
  paths: Array<Array<PathSegment>>
): Array<NodeValidationData> => {
  const ctx = useContext(Context);
  if (!ctx) throw new Error('No validation provider found. Make sure a <NodeValidationProvider /> is defined.')
  if (!ctx.aggregateNodeValidationHook) throw new Error(`Couldn't find a aggregateNodeValidationHook to use.`)

  return ctx.aggregateNodeValidationHook(paths);
}

export function NodeValidationProvider(
  {
    nodeValidationHook,
    aggregateNodeValidationHook,
    validatorRegistry,
    children
  }: ValidationProviderProps
) {
  return (
    <Context.Provider value={{
      nodeValidationHook,
      aggregateNodeValidationHook,
      validatorRegistry
    }}>
      {children}
    </Context.Provider>
  );
}