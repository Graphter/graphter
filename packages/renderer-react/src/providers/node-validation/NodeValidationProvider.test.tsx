import React from 'react'
import { NodeValidationProvider, useAggregateNodeValidation, useNodeValidation } from './NodeValidationProvider'
import { render } from "@testing-library/react";

describe('<NodeValidationProvider />', () => {
  it('should render children', () => {
    const { queryByText } = render(
      <NodeValidationProvider
        nodeValidationHook={jest.fn()}
        aggregateNodeValidationHook={jest.fn()}
        validatorRegistry={[]}
      >
        <div>Some child component</div>
      </NodeValidationProvider>
    )
    expect(queryByText('Some child component')).not.toBeNull()
  })
  describe('when using useNodeValidation hook', () => {
    function ValidationConsumerMock(props: any) {
      useNodeValidation(props.config, props.path)
      return null
    }
    it('should delegate to the registered node validation hook', () => {
      const nodeValidationHookMock = jest.fn()
      render(
        <NodeValidationProvider
          nodeValidationHook={nodeValidationHookMock}
          aggregateNodeValidationHook={jest.fn()}
          validatorRegistry={[]}
        >
          <ValidationConsumerMock path={['page']} config={{ id: 'some-id' }} />
        </NodeValidationProvider>
      )
      expect(nodeValidationHookMock).toHaveBeenCalledWith({ id: 'some-id' }, ['page'], [])
    })
    it('should throw an error if the provider is not present in the component hierarchy', () => {
      expect(() => {
        render(<ValidationConsumerMock path={['page']} />)
      }).toThrowErrorMatchingSnapshot()
    })
    it('should throw an error if a node validation hook has not been defined', () => {
      expect(() => {
        render(
          <NodeValidationProvider
            // @ts-ignore
            nodeValidationHook={null}
            aggregateNodeValidationHook={jest.fn()}
            validatorRegistry={[]}
          >
            <ValidationConsumerMock path={['page']} />
          </NodeValidationProvider>
        )
      }).toThrowErrorMatchingSnapshot()
    })
    it('should throw an error if a validator registry is not present', () => {
      expect(() => {
        render(
          <NodeValidationProvider
            nodeValidationHook={jest.fn()}
            aggregateNodeValidationHook={jest.fn()}
            // @ts-ignore
            validatorRegistry={null}
          >
            <ValidationConsumerMock path={['page']} />
          </NodeValidationProvider>
        )
      }).toThrowErrorMatchingSnapshot()
    })
  })
  describe('when using useAggregateNodeValidation', () => {
    function AggregateValidationConsumerMock(props: any) {
      useAggregateNodeValidation(props.paths)
      return null
    }
    it('should delegate to the registered aggregate node validation hook', () => {
      const aggregateNodeValidationHookMock = jest.fn()
      render(
        <NodeValidationProvider
          nodeValidationHook={jest.fn()}
          aggregateNodeValidationHook={aggregateNodeValidationHookMock}
          validatorRegistry={[]}
        >
          <AggregateValidationConsumerMock paths={[['page'], ['page', '0']]} />
        </NodeValidationProvider>
      )
      expect(aggregateNodeValidationHookMock).toHaveBeenCalledWith([['page'], ['page', '0']])
    })
    it('should throw an error if the provider is not present in the component hierarchy', () => {
      expect(() => {
        render(
          <AggregateValidationConsumerMock paths={[['page'], ['page', '0']]} />
        )
      }).toThrowErrorMatchingSnapshot()
    })
    it('should throw an error if an aggregate node validation hook has not been defined', () => {
      expect(() => {
        render(
          <NodeValidationProvider
            nodeValidationHook={jest.fn()}
            // @ts-ignore
            aggregateNodeValidationHook={null}
            validatorRegistry={[]}
          >
            <AggregateValidationConsumerMock paths={[['page'], ['page', '0']]} />
          </NodeValidationProvider>
        )
      }).toThrowErrorMatchingSnapshot()
    })
  })
})
export {}