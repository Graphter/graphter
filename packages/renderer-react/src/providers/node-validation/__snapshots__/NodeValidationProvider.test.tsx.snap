// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NodeValidationProvider /> when using useAggregateNodeValidation should throw an error if an aggregate node validation hook has not been defined 1`] = `"Couldn't find a aggregateNodeValidationHook to use."`;

exports[`<NodeValidationProvider /> when using useAggregateNodeValidation should throw an error if the provider is not present in the component hierarchy 1`] = `"No validation provider found. Make sure a <NodeValidationProvider /> is defined."`;

exports[`<NodeValidationProvider /> when using useNodeValidation hook should throw an error if a node validation hook has not been defined 1`] = `"Couldn't find a NodeValidationHook to use."`;

exports[`<NodeValidationProvider /> when using useNodeValidation hook should throw an error if a validator registry is not present 1`] = `"No validators have been defined. Please ensue <NodeValidationProvider /> has been defined correctly."`;

exports[`<NodeValidationProvider /> when using useNodeValidation hook should throw an error if the provider is not present in the component hierarchy 1`] = `"No validation provider found. Make sure a <NodeValidationProvider /> is defined."`;
